modules/grandpa/src/lib.rs

```diff
@@ -189,6 +189,7 @@ pub mod pallet {
 			ensure!(best_finalized_number < *number, <Error<T, I>>::OldHeader);
 
 			let authority_set = <CurrentAuthoritySet<T, I>>::get();
+			let extra_proof_size_bytes = authority_set.extra_proof_size_bytes();
```

Comment:
Sure - the `proof_size` takes into account all values that are read from runtime storage during the call. So we definitely want current value, not the new one

Reply:
When we talk about proof-size component of weights - yes. What it means (in brief): blockchain has deterministic state transition function, so we don't need proof of written values. What extrinsic writes is uniquely defined by the extrinsic code (which is a part of runtime code), current runtime state (runtime storage values) and the extrinsic arguments.

For example - let's consider the simplest balance transfer extrinsic - `transfer_to_new_account(payer, payee, amount)`. Let's say that it doesn't check that the `payee` account exists - it just inserts the `amount` into ledger under `payee` key. If we call it with arguments `transfer_to_new_account(Alice, Bob, 100)`, then it will make following storage operations (if it succeeds):
- read Alice balance;
- write updated Alice balance;
- write Bob balance.

This ^^^ will happen on our (bridge hub) parachain. But how do we prove to the relay chain (Kusama/Polkadot) that our block with the `transfer_to_new_account(Alice, Bob, 100)` call is correct? Relay chain has:
- the previous block header, which has the `state_root` field, which is a "digest" of all runtime storage right before new block starts executing;
- the parachain runtime code (afaik), so it can execute the extrinsic.

So relay chain may decode the parachain block extrinsic (`transfer_to_new_account(Alice, Bob, 100)`) and start executing it. But at some point, our extrinsic needs to read the Alice balance. There's where PoV gets into action. PoV is storage proof of all storage values that the extrinsic reads. So as you see ^^^ we have only read Alice balance and the PoV will contain some intermediate merkle trie nodes ++ leaf node with the Alice balance. The relay chain is able to verify that storage proof, because it has the `state_root` of the previous block.

Then, when it comes to the writes, you can see that the written values are defined by the code and by the previous state - we don't need to include them in PoV. Given old Alice balance, the relay chain sees that it needs to write updated Alice and Bob balances.

The final step is to check that the parachain has not cheated and has written correct values. That is simple - we have the `state_root` field in the new block and also we have the `state_root`, which we have computed on our own (using the old state and by emulating extrinsics). If they match, then the provided header is correct and relay chain accepts it.

So yes, we don't care about written values when we talk about PoV.

Reply:
Thanks for the detailed explanation ! It was very useful for me. I think I understand now.